pkg/commands/workload_apply.go

```diff
@@ -101,20 +120,40 @@ func (opts *WorkloadApplyOptions) Exec(ctx context.Context, c *cli.Config) error
 		}
 	}
 
-	workload.Name = opts.Name
-	workload.Namespace = opts.Namespace
-	if opts.FilePath != "" {
-		var serviceAccountCopy string
-		// avoid passing a nil pointer to MergeServiceAccountName func
-		if fileWorkload.Spec.ServiceAccountName != nil {
-			serviceAccountCopy = *fileWorkload.Spec.ServiceAccountName
-		}
+	if opts.UpdateStrategy == mergeUpdateStrategy {
+		if opts.FilePath != "" {
+			var serviceAccountCopy string
+			// avoid passing a nil pointer to MergeServiceAccountName func
+			if fileWorkload.Spec.ServiceAccountName != nil {
+				serviceAccountCopy = *fileWorkload.Spec.ServiceAccountName
+			}
 
-		workload.Spec.MergeServiceAccountName(serviceAccountCopy)
+			workload.Spec.MergeServiceAccountName(serviceAccountCopy)
+		}
+		workload.Merge(fileWorkload)
 	}
 
-	workload.Merge(fileWorkload)
+	if opts.UpdateStrategy == replaceUpdateStrategy {
+		// assign all the file workload fields to the workload in the cluster
+		workload = fileWorkload
+
+		// if there is a workload in the cluster with all metadata populated
+		// re assign the system populated fields so we won't find an error because of some missing fields
+		if currentWorkload != nil && !reflect.DeepEqual(currentWorkload.ObjectMeta, (metav1.ObjectMeta{})) {
```

Comment:
nit: create a helper func to move this logic and add unit tests for that

Reply:
I didn't do it as a separate func because there is no need to have a public func for that logic since it's only being used in apply and it's not likely to be used anywhere else.

Reply:
ok, separated logic and created unit tests